lib/beacon/loader/worker.ex

```diff
@@ -523,4 +563,26 @@ defmodule Beacon.Loader.Worker do
         result
     end
   end
+
+  # this is a global lock to ensure that `load_fn` only runs for one worker per module;
+  # duplicate workers will simply wait for the first one
+  defp safe_load(module, load_fn) do
+    case Registry.register(Beacon.Registry, module, module) do
+      {:ok, _pid} ->
+        # we are the first worker, let's do the work
+        load_fn.()
+        Registry.unregister(Beacon.Registry, module)
+        module
+
+      {:error, {:already_registered, pid}} ->
+        # another worker already started, let's wait for it
+        _ref = Process.monitor(pid)
+
+        receive do
+          {:DOWN, _ref, :process, _pid, {:shutdown, :loaded}} -> module
+        after
+          1_000 -> :error
+        end
+    end
+  end
```

Comment:
This is the main logic.  Tried it without the `unregister/2` at first and it had some issues in running tests, it would take forever to finish, I think due to the sandboxing.

So now it unregisters after compiling, and also lowered the timeout to 1_000